# Phase 1

## 1.1

### A

The input data consists of three `.csv` files containing information about patients, stations, and observations.

In [1]:
import os

data_path = './120/'

for f in os.listdir(data_path):
    print(f"{f}: {os.path.getsize(data_path + f)} Bytes")

patient.csv: 354346 Bytes
station.csv: 46128 Bytes
observation.csv: 4700498 Bytes


In [2]:
import pandas as pd

# load data from .csv files
delimiter = '\t'
patient = pd.read_csv(data_path + 'patient.csv', delimiter=delimiter)
station = pd.read_csv(data_path + 'station.csv', delimiter=delimiter)
observation = pd.read_csv(data_path + 'observation.csv', delimiter=delimiter)

File `patient.csv`: 

Size: 354,346 bytes \
Number of rows: 2124 \
Number of attributes: 13 \
Attributes:
| Attribute     | Type     |
|----------------|----------|
| username       | object   |
| mail           | object   |
| job            | object   |
| address        | object   |
| registration   | object   |
| name           | object   |
| blood_group    | object   |
| residence      | float64  |
| birthdate      | object   |
| user_id        | int64    |
| ssn            | object   |
| company        | object   |
| station_ID     | int64    |

In [3]:
patient.info()
patient.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2124 entries, 0 to 2123
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   username      2124 non-null   object 
 1   mail          2124 non-null   object 
 2   job           637 non-null    object 
 3   address       1805 non-null   object 
 4   registration  2124 non-null   object 
 5   name          2124 non-null   object 
 6   blood_group   2124 non-null   object 
 7   residence     0 non-null      float64
 8   birthdate     1168 non-null   object 
 9   user_id       2124 non-null   int64  
 10  ssn           2124 non-null   object 
 11  company       2124 non-null   object 
 12  station_ID    2124 non-null   int64  
dtypes: float64(1), int64(2), object(10)
memory usage: 215.8+ KB


username                         mail  job  \
0      paulina59         vincenza05@libero.it  NaN   
1      unakamura         imatsumoto@gmail.com  NaN   
2  sandrogarzoni  spallanzanisilvestro@tin.it  NaN   
3       hanako08    naokoishikawa@hotmail.com  NaN   
4         ajones        bauergloria@yahoo.com  NaN   

                                             address          registration  \
0          Rotonda Matilda, 169\n57014, Crocino (LI)            2024-12-02   
1                      兵庫県川崎市高津区箭坪8丁目22番17号 橋場ハイツ987           19 Aug 2024   
2  Borgo Patrizio, 18 Appartamento 81\n43040, Var...           11 Mar 2025   
3                                 福島県港区南赤田42丁目27番17号  02/20/2020, 00:00:00   
4   800 Jennifer Hills\nNew Christianville, SD 66147           20 May 2020   

                       name blood_group  residence   birthdate  user_id  \
0  Bruno Legnante-Pincherle         AB-        NaN         NaN      634   
1                     高橋 里佳          A+        NaN  1994-09-02     1577   
2         Stella Zaccagnini          O+        NaN         NaN      360   
3                     山崎 桃子         AB+        NaN         NaN      496   
4             Bethany Joyce         AB+        NaN  1953-11-10      960   

                ssn                 company  station_ID  
0  CRNTTE88H08D352G  Mengolo-Vivaldi s.r.l.         607  
1       631-24-8857                近藤電気合同会社          69  
2  RFFFLP44A09G036G         Guariento Group          25  
3       492-47-8815               長谷川銀行合同会社         365  
4       686-81-7668    Fitzpatrick and Sons         156

File `station.csv`:

Size: 46,128 Bytes \
Number of rows: 776 \
Number of attributes: 6 \
Attributes:
| Attribute | Type     |
|------------|----------|
| longitude  | float64  |
| location   | object   |
| code       | object   |
| revision   | object   |
| station    | object   |
| latitude   | float64  |

In [4]:
station.info()
station.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 776 entries, 0 to 775
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   longitude  776 non-null    float64
 1   location   776 non-null    object 
 2   code       776 non-null    object 
 3   revision   776 non-null    object 
 4   station    776 non-null    object 
 5   latitude   776 non-null    float64
dtypes: float64(2), object(4)
memory usage: 36.5+ KB


longitude          location code     revision     station  latitude
0  -74.28459  America/New_York   US   2024/10/10  Woodbridge  40.55760
1    7.11393     Europe/Berlin   DE   2024-07-30   Burscheid  51.08468
2   84.63445      Asia/Kolkata   IN  29 Jan 2023    Rafiganj  24.81757
3   -8.53778   Africa/Monrovia   LR  13 Aug 2018  New Yekepa   7.57944
4   -5.53715     Europe/London   GB  13 Jan 2018    Penzance  50.11861

File `observation.csv`:

Size: 4,700,498 Bytes \
Number of rows: 10800 \
Number of attributes: 23 \
Attributes:
| Attribute               | Type     |
|--------------------------|----------|
| SpO₂                     | float64  |
| HR                       | float64  |
| PI                       | float64  |
| RR                       | float64  |
| EtCO₂                    | float64  |
| FiO₂                     | float64  |
| PRV                      | float64  |
| BP                       | float64  |
| Skin Temperature         | float64  |
| Motion/Activity index    | float64  |
| PVI                      | float64  |
| Hb level                 | float64  |
| SV                       | float64  |
| CO                       | float64  |
| Blood Flow Index         | float64  |
| PPG waveform features    | float64  |
| Signal Quality Index     | float64  |
| Respiratory effort       | float64  |
| O₂ extraction ratio      | float64  |
| SNR                      | float64  |
| oximetry                 | float64  |
| latitude                 | float64  |
| longitude                | float64  |

In [5]:
observation.info()
observation.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10800 entries, 0 to 10799
Data columns (total 23 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   SpO₂                   10800 non-null  float64
 1   HR                     10800 non-null  float64
 2   PI                     10800 non-null  float64
 3   RR                     10800 non-null  float64
 4   EtCO₂                  10800 non-null  float64
 5   FiO₂                   10800 non-null  float64
 6   PRV                    10800 non-null  float64
 7   BP                     10800 non-null  float64
 8   Skin Temperature       10800 non-null  float64
 9   Motion/Activity index  10800 non-null  float64
 10  PVI                    10800 non-null  float64
 11  Hb level               10800 non-null  float64
 12  SV                     10800 non-null  float64
 13  CO                     10800 non-null  float64
 14  Blood Flow Index       10800 non-null  float64
 15  PP

SpO₂         HR         PI         RR      EtCO₂       FiO₂  \
0  98.057687  81.561900   9.634841  15.794092  38.357063  70.144098   
1  98.237816  91.922786  13.519149  17.994040  36.957369  68.604381   
2  97.577230  85.449842  10.051098  15.246586  38.209241  55.653927   
3  97.216777  75.182277  11.845941  15.411316  38.034613  58.892038   
4  97.612658  90.212598  11.042576  16.769490  37.190540  54.631574   

          PRV          BP  Skin Temperature  Motion/Activity index  ...  \
0   92.163941   99.321851         35.916763               7.798693  ...   
1   78.197845  103.692629         34.995809               8.316729  ...   
2  126.801787  101.691015         35.523631               8.979466  ...   
3  140.348222  104.877599         35.750361               8.149849  ...   
4   78.477997  109.071771         36.037151               7.784455  ...   

         CO  Blood Flow Index  PPG waveform features  Signal Quality Index  \
0  4.002487         59.928273              55.124002             62.925756   
1  4.143390         75.075702              39.349704             42.434930   
2  4.010376         35.361168              39.482841             56.390250   
3  4.000762         29.386672              41.344283             62.346638   
4  4.070999         38.118630              85.759457             47.176929   

   Respiratory effort  O₂ extraction ratio        SNR  oximetry  latitude  \
0           33.555239             0.241442  36.157403       0.0  33.76446   
1           55.008367             0.260622  28.446366       1.0  52.45810   
2           74.881875             0.206879  27.984798       1.0  44.92934   
3           69.884494             0.224859  37.210683       0.0  43.32393   
4           51.488496             0.208957  36.468848       1.0  42.16808   

   longitude  
0 -117.79394  
1   13.28702  
2   37.99117  
3    5.45840  
4  -88.42814  

[5 rows x 23 columns]

Description of sensor variables:

In [6]:
sensor_variables = pd.read_csv('sensor_variable_range.csv', delimiter='	')

sensor_variables

Variable         Unit  \
0                    SpO₂            %   
1                      HR          bpm   
2                      PI            %   
3                      RR  breaths/min   
4                   EtCO₂         mmHg   
5                    FiO₂            %   
6                     PRV           ms   
7                      BP         mmHg   
8        Skin Temperature           °C   
9   Motion/Activity index            -   
10                    PVI            %   
11               Hb level         g/dL   
12                     SV      mL/beat   
13                     CO        L/min   
14       Blood Flow Index            -   
15  PPG waveform features            -   
16   Signal Quality Index            -   
17     Respiratory effort            -   
18    O₂ extraction ratio            -   
19                    SNR           dB   

                                     Description        Value Range  
0                   Peripheral oxygen saturation            95–100%  
1                 Heart rate from pulse oximeter         60–100 bpm  
2                                Perfusion index            0.2–20%  
3                               Respiratory rate  12–20 breaths/min  
4                       End-tidal carbon dioxide         35–45 mmHg  
5                    Fraction of inspired oxygen            21–100%  
6                         Pulse rate variability          20–200 ms  
7       Blood pressure (systolic diastolic mean)  90/60–120/80 mmHg  
8                         Peripheral temperature           33–38 °C  
9                     Sensor artifact correction                NaN  
10                       Pleth Variability Index             10–20%  
11                      Hemoglobin concentration         12–18 g/dL  
12                                 Stroke volume     60–100 mL/beat  
13                                Cardiac output          4–8 L/min  
14  Optical/near-infrared blood flow measurement     Relative units  
15        Photoplethysmography signal parameters                NaN  
16             Reliability indicator of readings              0–100  
17        Chest impedance or belt-derived effort                NaN  
18           Calculated oxygen extraction metric            0.2–0.3  
19             Signal-to-noise ratio from sensor           20–40 dB

### B

### C

### D

### E

## 1.2

### A

In [7]:
def format_dates(date_to_format):
    # 'coerce' - will replace undefined data formats as NaT (Not a Time)
    # 'mixed'  - try to infer the format for each value individually
    return pd.to_datetime(date_to_format, format='mixed', errors='coerce')

#### patient

Problem 1 - Inconsistent date formats

In [8]:
patient['registration'] = format_dates(patient['registration'])
patient['birthdate'] = format_dates(patient['birthdate'])

Potential problem - Invalid or irrelevant values of blood type

In [9]:
blood_group_values = patient['blood_group'].unique()
print(sorted(blood_group_values))
print(f"Number of null/NaN blood group type values: {patient['blood_group'].isnull().sum()}")

### Everything is alright.

['A+', 'A-', 'AB+', 'AB-', 'B+', 'B-', 'O+', 'O-']
Number of null/NaN blood group type values: 0


Problem 2 - Date of birth later than date of registration

In [10]:
invalid_records = patient[patient['registration'] < patient['birthdate']]
print(f"Number of invalid records (birthdate < registration date): {len(invalid_records)}")

patient.drop(index=invalid_records.index, inplace=True)

Number of invalid records (birthdate < registration date): 44


Potential problem - User duplicates (user_id, station_ID)

In [11]:
duplicates = patient[patient[['user_id','station_ID']].duplicated(keep=False)]
print(f"Number of invalid records (user_id duplicates): {len(duplicates)}")

patient.drop(index=duplicates.index, inplace=True)

Number of invalid records (user_id duplicates): 0


Potential problem - Missed values

In [12]:
patient_misses = patient.isnull().sum()
print(f"Number of missing values per attribute in `patient` table: \n{patient_misses}")

Number of missing values per attribute in `patient` table: 
username           0
mail               0
job             1452
address          313
registration       0
name               0
blood_group        0
residence       2080
birthdate        956
user_id            0
ssn                0
company            0
station_ID         0
dtype: int64


#### station

Problem 1 - Inconsistent date formats

In [13]:
station['revision'] = format_dates(station['revision'])

Potential problem - Incorrect/Missing latitude and/or longitude values

In [14]:
wrong_latitude = station[(station['latitude'] < -90) | (station['latitude'] > 90)]
wrong_longitude = station[(station['latitude'] < -180) | (station['latitude'] > 180)]
missing_coordinates = station[station['latitude'].isna() | station['longitude'].isna()]
print(f"Number of records with incorrect latitude: {len(wrong_latitude)}")
print(f"Number of records with incorrect longitude: {len(wrong_longitude)}")
print(f"Number of records with missing coordinates: {len(missing_coordinates)}")

Number of records with incorrect latitude: 0
Number of records with incorrect longitude: 0
Number of records with missing coordinates: 0


Potential problem - Duplicates

In [15]:
station_duplicates = station.duplicated().sum()
print(f"Number of duplicated records: {station_duplicates}")

Number of duplicated records: 0


Potential problem - Missed values

In [16]:
station_misses = station.isnull().sum()
print(f"Number of missing values per attribute in `station` table: \n{station_misses}")

Number of missing values per attribute in `station` table: 
longitude    0
location     0
code         0
revision     0
station      0
latitude     0
dtype: int64


#### observation

Potential problem - Wrong data type (invalid or inconsistent entries)

In [17]:
observation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10800 entries, 0 to 10799
Data columns (total 23 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   SpO₂                   10800 non-null  float64
 1   HR                     10800 non-null  float64
 2   PI                     10800 non-null  float64
 3   RR                     10800 non-null  float64
 4   EtCO₂                  10800 non-null  float64
 5   FiO₂                   10800 non-null  float64
 6   PRV                    10800 non-null  float64
 7   BP                     10800 non-null  float64
 8   Skin Temperature       10800 non-null  float64
 9   Motion/Activity index  10800 non-null  float64
 10  PVI                    10800 non-null  float64
 11  Hb level               10800 non-null  float64
 12  SV                     10800 non-null  float64
 13  CO                     10800 non-null  float64
 14  Blood Flow Index       10800 non-null  float64
 15  PP

Potential problem - Incorrect/Missing latitude and/or longitude values

In [18]:
wrong_latitude = observation[(observation['latitude'] < -90) | (observation['latitude'] > 90)]
wrong_longitude = observation[(observation['latitude'] < -180) | (observation['latitude'] > 180)]
missing_coordinates = observation[observation['latitude'].isna() | observation['longitude'].isna()]
print(f"Number of records with incorrect latitude: {len(wrong_latitude)}")
print(f"Number of records with incorrect longitude: {len(wrong_longitude)}")
print(f"Number of records with missing coordinates: {len(missing_coordinates)}")

Number of records with incorrect latitude: 0
Number of records with incorrect longitude: 0
Number of records with missing coordinates: 1


Potential problem - Duplicates

In [19]:
observation_duplicates = observation.duplicated().sum()
print(f"Number of duplicated records: {observation_duplicates}")

Number of duplicated records: 0


Potential problem - Missed values

In [20]:
observation_misses = observation.isnull().sum()
print(f"Number of missing values per attribute in `observation` table: \n{observation_misses}")

Number of missing values per attribute in `observation` table: 
SpO₂                     0
HR                       0
PI                       0
RR                       0
EtCO₂                    0
FiO₂                     0
PRV                      0
BP                       0
Skin Temperature         0
Motion/Activity index    0
PVI                      0
Hb level                 0
SV                       0
CO                       0
Blood Flow Index         0
PPG waveform features    0
Signal Quality Index     0
Respiratory effort       0
O₂ extraction ratio      0
SNR                      0
oximetry                 0
latitude                 0
longitude                1
dtype: int64


### B

In [ ]:
def get_range():
    str_range = sensor_variables.apply(lambda row: str(row['Value Range']).replace(str(row['Unit']), ''), axis=1)
    str_range = str_range.apply(lambda row: row.strip())
    str_range = str_range.to_list()
    str_range = [rng if rng not in ['nan', 'Relative units'] else '' for rng in str_range]

    value_range = []
    for rng in str_range:
        if rng == '':
            value_range.append([None, None])
        else:
            range = rng.split('–')
            range = [range[0].split('/')[0], range[1].split('/')[0]]
            value_range.append(range)

    sensor_variables['min_value'] = [float(x[0]) if x[0] is not None else None for x in value_range]
    sensor_variables['max_value'] = [float(x[1]) if x[1] is not None else None for x in value_range]

get_range()
sensor_variables

# get list of attributes
attributes = [attr for attr in observation.columns if attr in sensor_variables['Variable'].values]

# empty DataFrame for counting violations
out_of_range = pd.DataFrame(False, index=observation.index, columns=attributes)

for attr in attributes:
    min = sensor_variables.loc[sensor_variables['Variable'] == attr, 'min_value'].values[0]
    max = sensor_variables.loc[sensor_variables['Variable'] == attr, 'max_value'].values[0]
    
    for idx, val in observation[attr].items():
        if pd.isna(val):
            continue  # NaN in value skip
        if pd.isna(min) or pd.isna(max):
            continue  # NaN in range skip
        if val < min or val > max:
            out_of_range.at[idx, attr] = True
# count number of violations per attribute
violations_count = out_of_range.sum()
violations_count

SpO₂                     0
HR                       0
PI                       0
RR                       0
EtCO₂                    0
FiO₂                     0
PRV                      0
BP                       0
Skin Temperature         0
Motion/Activity index    0
PVI                      0
Hb level                 0
SV                       0
CO                       0
Blood Flow Index         0
PPG waveform features    0
Signal Quality Index     0
Respiratory effort       0
O₂ extraction ratio      0
SNR                      0
dtype: int64

In [22]:
def get_outliers_iqr(input_observation):
    outliers_iqr = pd.DataFrame(False, index=input_observation.index, columns=attributes)

    for attr in attributes:
        Q1 = input_observation[attr].quantile(0.25)
        Q3 = input_observation[attr].quantile(0.75)
        IQR = Q3 - Q1
        lower = Q1 - 1.5 * IQR
        upper = Q3 + 1.5 * IQR
        
        outliers_iqr[attr] = (input_observation[attr] < lower) | (input_observation[attr] > upper)
    return outliers_iqr

get_outliers_iqr(observation).sum()

SpO₂                       93
HR                          0
PI                         14
RR                          0
EtCO₂                       0
FiO₂                       69
PRV                        71
BP                         71
Skin Temperature          112
Motion/Activity index      88
PVI                        58
Hb level                  129
SV                         66
CO                       1429
Blood Flow Index           72
PPG waveform features      79
Signal Quality Index       64
Respiratory effort         98
O₂ extraction ratio         0
SNR                         0
dtype: int64

In [ ]:
#observation_no_outliers = observation[~get_outliers_iqr(observation).any(axis=1)] # one try is not enough

def remove_outliers(input_observation):
    observation_no_outliers = input_observation.copy()

    while True:
        if int(get_outliers_iqr(observation_no_outliers).sum().sum()) == 0:
            break

        observation_no_outliers = observation_no_outliers[~get_outliers_iqr(observation_no_outliers).any(axis=1)]

    return observation_no_outliers

# get_outliers_iqr(observation_no_outliers).sum()
observation_no_outliers = remove_outliers(observation)
get_outliers_iqr(observation_no_outliers).sum()
abs(len(observation_no_outliers) - len(observation))
#len(observation_no_outliers)
# observation_no_outliers['']

2992

In [78]:
def replace_outliers(input_observation):
    observation_replaced_outliers = input_observation.copy()
    bias = 0.14

    for attr in attributes:
        lower = input_observation[attr].quantile(bias)
        upper = input_observation[attr].quantile(1.00 - bias)

        observation_replaced_outliers[attr] = observation_replaced_outliers[attr].mask(input_observation[attr] < lower, lower)
        observation_replaced_outliers[attr] = observation_replaced_outliers[attr].mask(input_observation[attr] > upper, upper)
    return observation_replaced_outliers
#get_outliers_iqr(observation).sum()
get_outliers_iqr(replace_outliers(observation)).sum()

SpO₂                     0
HR                       0
PI                       0
RR                       0
EtCO₂                    0
FiO₂                     0
PRV                      0
BP                       0
Skin Temperature         0
Motion/Activity index    0
PVI                      0
Hb level                 0
SV                       0
CO                       0
Blood Flow Index         0
PPG waveform features    0
Signal Quality Index     0
Respiratory effort       0
O₂ extraction ratio      0
SNR                      0
dtype: int64

### C